py4DSTEM/process/phase/parameter_optimize.py

```diff
@@ -0,0 +1,449 @@
+            Keyword arguments passed to the preprocess method the reconstruction object
+        reconstruction_args: dict
+            Keyword arguments passed to the reconstruct method the the reconstruction object
+        affine_args: dict
+            Keyword arguments passed to AffineTransform. The transform is applied to the initial
+            scan positions.
+        """
+
+        # loop over each argument dictionary and split into static and optimization variables
+        (
+            self._init_static_args,
+            self._init_optimize_args,
+        ) = self._split_static_and_optimization_vars(init_args)
+        (
+            self._affine_static_args,
+            self._affine_optimize_args,
+        ) = self._split_static_and_optimization_vars(affine_args)
+        (
+            self._preprocess_static_args,
+            self._preprocess_optimize_args,
+        ) = self._split_static_and_optimization_vars(preprocess_args)
+        (
+            self._reconstruction_static_args,
+            self._reconstruction_optimize_args,
+        ) = self._split_static_and_optimization_vars(reconstruction_args)
+
+        # Save list of skopt parameter objects and inital guess
+        self._parameter_list = []
+        self._x0 = []
+        for k, v in (
+            self._init_optimize_args
+            | self._affine_optimize_args
+            | self._preprocess_optimize_args
+            | self._reconstruction_optimize_args
+        ).items():
+            self._parameter_list.append(v._get(k))
+            self._x0.append(v._initial_value)
+
+        self._init_args = init_args
+        self._affine_args = affine_args
+        self._preprocess_args = preprocess_args
+        self._reconstruction_args = reconstruction_args
+
+        self._reconstruction_type = reconstruction_type
+
+        self._set_optimizer_defaults()
+
+    def optimize(
+        self,
+        n_calls: int,
+        n_initial_points: int,
+        **skopt_kwargs: dict,
+    ):
+        """
+        Run optimizer
+
+        Parameters
+        ----------
+        n_calls: int
+            Number of times to run ptychographic reconstruction
+        n_initial_points: int
+            Number of uniformly spaced trial points to test before
+            beginning Bayesian optimization (must be less than n_calls)
+        skopt_kwargs: dict
+            Additional arguments to be passed to skopt.gp_minimize
+
+        """
+        self._optimization_function = self._get_optimization_function(
+            self._reconstruction_type,
+            self._parameter_list,
+            n_calls,
+            self._init_static_args,
+            self._affine_static_args,
+            self._preprocess_static_args,
+            self._reconstruction_static_args,
+            self._init_optimize_args,
+            self._affine_optimize_args,
+            self._preprocess_optimize_args,
+            self._reconstruction_optimize_args,
+        )
```

Comment:
Error divergence check and log error suggestion (see below):

```suggestion
            check_error_divergence = check_error_divergence,
            return_log_error = return_log_error,
        )
```